server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingService.java

```diff
@@ -0,0 +1,105 @@
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.admin.cluster.shards.routing.weighted.put.ClusterPutWeightedRoutingRequest;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.ClusterStateUpdateTask;
+import org.opensearch.cluster.ack.ClusterStateUpdateResponse;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.metadata.WeightedRoutingMetadata;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.Priority;
+import org.opensearch.common.component.AbstractLifecycleComponent;
+import org.opensearch.common.inject.Inject;
+
+import org.opensearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+
+/**
+ * * Service responsible for updating cluster state metadata with weighted routing weights
+ */
+public class WeightedRoutingService extends AbstractLifecycleComponent {
+    private static final Logger logger = LogManager.getLogger(WeightedRoutingService.class);
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+
+    @Inject
+    public WeightedRoutingService(ClusterService clusterService, ThreadPool threadPool) {
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+    }
+
+    public void registerWeightedRoutingMetadata(
+        final ClusterPutWeightedRoutingRequest request,
+        final ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        final WeightedRoutingMetadata newWeightedRoutingMetadata = new WeightedRoutingMetadata(request.getWeightedRouting());
+        clusterService.submitStateUpdateTask("update_weighted_routing", new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                Metadata metadata = currentState.metadata();
+                Metadata.Builder mdBuilder = Metadata.builder(currentState.metadata());
+                WeightedRoutingMetadata weightedRoutingMetadata = metadata.custom(WeightedRoutingMetadata.TYPE);
+                if (weightedRoutingMetadata == null) {
+                    logger.info("put weighted routing weights in metadata [{}]", request.getWeightedRouting());
+                    weightedRoutingMetadata = new WeightedRoutingMetadata(request.getWeightedRouting());
+                } else {
+                    WeightedRoutingMetadata changedMetadata = new WeightedRoutingMetadata(new WeightedRouting(null, null));
+                    if (!checkIfSameWeightsInMetadata(newWeightedRoutingMetadata, weightedRoutingMetadata)) {
+                        logger.info("updated weighted routing weights [{}] in metadata", request.getWeightedRouting());
+                        changedMetadata.setWeightedRouting(newWeightedRoutingMetadata.getWeightedRouting());
```

Comment:
Do we really need `changedMetadata`? Can we replace this line with:
```
weightedRoutingMetadata = new WeightedRoutingMetadata(newWeightedRoutingMetadata.getWeightedRouting());
```

and remove line 66?

Reply:
addressed